File: chat-getting-started-dropwizard/src/main/java/io/axoniq/labs/chat/commandmodel/ChatRoom.java
package io.axoniq.labs.chat.commandmodel;

import io.axoniq.labs.chat.coreapi.CreateRoomCommand;
import io.axoniq.labs.chat.coreapi.JoinRoomCommand;
import io.axoniq.labs.chat.coreapi.LeaveRoomCommand;
import io.axoniq.labs.chat.coreapi.MessagePostedEvent;
import io.axoniq.labs.chat.coreapi.ParticipantJoinedRoomEvent;
import io.axoniq.labs.chat.coreapi.ParticipantLeftRoomEvent;
import io.axoniq.labs.chat.coreapi.PostMessageCommand;
import io.axoniq.labs.chat.coreapi.RoomCreatedEvent;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.modelling.command.AggregateIdentifier;

import java.util.HashSet;
import java.util.Set;

import static org.axonframework.modelling.command.AggregateLifecycle.apply;

public class ChatRoom {

    @AggregateIdentifier
    String roomId;

    Set<String> participants = new HashSet<>();

    ChatRoom() {

    }

    @CommandHandler
    public ChatRoom(CreateRoomCommand command) {
        apply(new RoomCreatedEvent(command.getRoomId(), command.getName()));
    }

    @CommandHandler
    public void handle(JoinRoomCommand command) {
        if (!hasJoined(command.getParticipant())) {
            apply(new ParticipantJoinedRoomEvent(command.getRoomId(), command.getParticipant()));
        }
    }

    @CommandHandler
    public void handle(LeaveRoomCommand command) {
        if (hasJoined(command.getParticipant())) {
            apply(new ParticipantLeftRoomEvent(command.getRoomId(), command.getParticipant()));
        }
    }

    @CommandHandler
    public void handle(PostMessageCommand command) {
        if (!hasJoined(command.getParticipant())) {
            throw new IllegalStateException("You are not part of this room.");
        }
        apply(new MessagePostedEvent(command.getRoomId(), command.getParticipant(), command.getMessage()));
    }

    private boolean hasJoined(String participant) {
        return participants.contains(participant);
    }

    @EventSourcingHandler
    public void on(RoomCreatedEvent event) {
        this.roomId = event.getRoomId();
    }

    @EventSourcingHandler
    public void on(ParticipantJoinedRoomEvent event) {
        participants.add(event.getParticipant());
    }

    @EventSourcingHandler
    public void on(ParticipantLeftRoomEvent event) {
        participants.remove(event.getParticipant());
    }
}

File: chat-getting-started-kotlin/src/main/kotlin/io/axoniq/labs/chat/commandmodel/ChatRoom.kt
package io.axoniq.labs.chat.commandmodel

import io.axoniq.labs.chat.coreapi.*
import org.axonframework.commandhandling.CommandHandler
import org.axonframework.eventsourcing.EventSourcingHandler
import org.axonframework.modelling.command.AggregateIdentifier
import org.axonframework.modelling.command.AggregateLifecycle
import org.axonframework.spring.stereotype.Aggregate

@Aggregate
class ChatRoom {

    @AggregateIdentifier
    private var roomId: String? = null
    private var participants: MutableSet<String> = mutableSetOf()

    constructor() {} // required by Axon

    @CommandHandler
    constructor(command: CreateRoomCommand) {
        AggregateLifecycle.apply(RoomCreatedEvent(command.roomId, command.name))
    }

    @CommandHandler
    fun handle(command: JoinRoomCommand) {
        if (!participants.contains(command.participant)) {
            AggregateLifecycle.apply(ParticipantJoinedRoomEvent(roomId!!, command.participant))
        }
    }

    @CommandHandler
    fun handle(command: LeaveRoomCommand) {
        if (participants.contains(command.participant)) {
            AggregateLifecycle.apply(ParticipantLeftRoomEvent(roomId!!, command.participant))
        }
    }

    @CommandHandler
    fun handle(command: PostMessageCommand) {
        check(participants.contains(command.participant)) { "You cannot post messages unless you've joined the chat room" }
        AggregateLifecycle.apply(
                MessagePostedEvent(roomId!!, command.participant, command.message)
        )
    }

    @EventSourcingHandler
    protected fun on(event: RoomCreatedEvent) {
        roomId = event.roomId
    }

    @EventSourcingHandler
    protected fun on(event: ParticipantJoinedRoomEvent) {
        participants.add(event.participant)
    }

    @EventSourcingHandler
    protected fun on(event: ParticipantLeftRoomEvent) {
        participants.remove(event.participant)
    }
}
